shared/hexDocumentModel.ts

```diff
@@ -0,0 +1,320 @@
+		if (!this._edits.some(e => e.op !== HexDocumentEditOp.Replace) && !this.readAccessor) {
+			await this.accessor.writeBulk(edits.map(e => ({ offset: e.offset, data: (e as HexDocumentReplaceEdit).value })));
+		} else {
+			await this.accessor.writeStream(this.readUsingRanges(ranges, 0));
+		}
+	}
+
+	/**
+	 * Revents to the contents last saved on disk, discarding edits,
+	 */
+	public revert(): void {
+		this._edits = [];
+		this._editTimeline = undefined;
+		this.generation++;
+	}
+
+	/**
+	 * Applies edits to the model, reutnring the handle. Undo *must* only be
+	 * called once all subsequently made edits have also been undone, and
+	 * vise versa for `redo`.
+	 */
+	public makeEdits(edits: HexDocumentEdit[]): HexDocumentEditReference {
+		let currentGen = this.generation;
+		let currentIndex = this._edits.length;
+		this._edits.push(...edits);
+		this._editTimeline = undefined;
+
+		return {
+			undo: () => {
+				if (this.generation === currentGen) {
+					this._edits.splice(currentIndex, edits.length);
+				} else {
+					this._edits.push(...edits.map(reverseEdit));
+				}
+				this._editTimeline = undefined;
+			},
+			redo: () => {
+				currentIndex = this._edits.length;
+				currentGen = this.generation;
+				this._edits.push(...edits);
+				this._editTimeline = undefined;
+			},
+		};
+	}
+
+	private async *readUsingRanges(ranges: readonly Range[], fromOffset: number, chunkSize = 1024): AsyncIterableIterator<Uint8Array>	{
+		const buf = new Uint8Array(chunkSize);
+		const readAccessor = this.readAccessor || this.accessor;
+		for (let i = 0; i < ranges.length; i++) {
+			const range = ranges[i];
+			if (range.op === RangeOp.Skip) {
+				continue;
+			}
+
+			if (range.op === RangeOp.Insert) {
+				const offset = Math.max(0, fromOffset - range.offset);
+				let toYield = range.value;
+				if (offset < range.value.length && offset > 0) {
+					toYield = range.value.subarray(offset);
+				}
+				if (toYield.length > 0) {
+					yield toYield;
+				}
+				continue;
+			}
+
+			// range.op === Read
+			const until = i + 1 < ranges.length ? ranges[i + 1].offset - range.offset : Infinity;
+			let roffset = range.roffset + Math.max(0, fromOffset - range.offset);
+			while (roffset < until) {
+				const bytes = await readAccessor.read(roffset, buf.subarray(0, Math.min(buf.length, until - roffset)));
+				if (bytes === 0) {
+					break; // EOF
+				}
+				yield buf.subarray(0, bytes);
+				roffset += bytes;
+			}
+		}
+	}
+
+	/**
+	 * Gets the size of the file on disk. Takes the current generation, so that
+	 * the file is re-read whenever the generation changes.
+	 */
+	private readonly getSizeInner =
+		memoizeLast((_generation: number) => (this.readAccessor || this.accessor).getSize());
+
+	/**
+	 * Gets instructions for returning file data. Returns a list of contiguous
+	 * `Range` instances. Each "range" issues instructions until the next
+	 * "offset". This can be used to generate file data.
+	 */
+	private getEditTimeline() {
```

Comment:
note: today we invalidate the entire timeline when an edit is made, but in theory we could just compute that onto the existing timeline. However I expect this to function to be fast enough as to make no difference